### PR TITLE
Adapt to Mirage-crypto 0.8.1

### PR DESCRIPTION
**Pull request by anmonteiro**
This adapts the Jwe module to the new interfaces in mirage-crypto [0.8.1](https://github.com/mirage/mirage-crypto/releases/tag/v0.8.1).

This is still WIP and the following are unresolved:

- [ ] I'd appreciate some guidance on what to do about the auth tag for `A256GCM` (cc @hannesm)
- [x] Needs updating the esy file + lockfiles but I don't run esy so someone feel free to push to this PR

### CODE REVIEW

**Comment by ulrikstrid** [comment]
Thanks a lot for this.

I can fix the esy.lock when it's done. 👍

File: dune-project [hunk 1]
```diff
@@ -19,7 +19,7 @@
    (ocaml (>= 4.07.0))
```
Comment (ulrikstrid) [inline dune-project:19]: Does this need to be bumped as well?

**Comment by hannesm** [comment]
hi, this looks sensible. sorry for the breakage, I was not aware that JOSE uses the tag separately. does JOSE specify other AEAD ciphers than AES+GCM (such as AES+CCM)? (if so, https://github.com/mirage/mirage-crypto/issues/74 strives for a unified interface that may reduce duplication). it may as well be worth to introduce another set of decrypt + encrypt functions which are similar to the old ones (i.e. return the tag and ciphertext separately) - though when reading AEAD, decode should always verify the tag (instead of leaving it to the caller).

the reason for the change in AES-GCM is that all the protocols I am aware of (tls, ssh) transmit the encrypted data followed directly by the tag, thus the append (`encrypt`) or split (`decrypt`) was duplicated on the use side.

**Comment by ulrikstrid** [comment]
I'm okay with dropping OCaml <4.8.0, I'm going to go ahead and merge this and then I'll release a 0.5.0 as soon as I can find some spare time.
Thanks a lot @anmonteiro and thanks @hannesm for commenting on this.

We currently only have AES+GCM but I think AES+CCM is valid in the JOSE spec, I'm in no rush to add more algorithms currently since I'm not using the JWE part myself yet.

**Pull request merged**